packages/arb-token-bridge-ui/src/components/App/App.tsx

```diff
@@ -172,21 +168,75 @@ const Injector = ({ children }: { children: React.ReactNode }): JSX.Element => {
       const l1Address = await l1Signer.getAddress()
       const l2Address = await l2Signer.getAddress()
 
-      try {
-        const l1AddressIsEOA = await addressIsEOA(l1Address, l1Provider)
-        const l2AddressIsEOA = await addressIsEOA(l2Address, l2Provider)
+      const walletType: WalletType = await (async () => {
+        const l1AddressIsSmartContract = !(await addressIsEOA(
+          l1Address,
+          l1Provider
+        ))
+        const l2AddressIsSmartContract = !(await addressIsEOA(
+          l2Address,
+          l2Provider
+        ))
+
+        if (!l1AddressIsSmartContract && !l2AddressIsSmartContract) {
+          return WalletType.EOA
+        }
+
+        // TODO: Remove this comment after questions answered!
+        // QUESTION 1: Is this true that only 1 address is required to be smart contract?
+        // Since destination address is specified my reasoning is it's enough for just
+        // L1 to be SC for deposits, and L2 for withdrawals.
```

Comment:
So the key thing that this PR adds is ability for a user to input their own destination address. Given that, we can simplify things: I don't think we need any special case handling for gnosis wallets, and I don't think we need any notion of an "UNSUPPORTED_CONTRACT_WALLET" at all.

The heuristic can now be as follows — say user has their wallet connected to address 0xabc:

- If 0xabc is an EOA on both L1 and L2, we set the destination address to 0xabc by default (as we do now on the live site); changing the destination address can be offered in "advanced settings" mode.
- If 0xabc is a contract on L1 or L2 (or both), we _force_ the user to manually input their own destination address; i.e., no default is provided. 

Anytime a user manually inputs a destination address, the UI probably should (at some point in the flow) tell them whether the address they inputed is an EOA or a contract, just as a sanity check. 

I may well be out of he loop on some contextual UX discussions so don't immediately take what I'm saying as gospel, but that's my view

Reply:
@DZGoldman what if there are 2 smart contract addresses with mismatching owners? Wouldn't that count as an unsupported/invalid SC wallet?